### PR TITLE
🔒 Warden: Hardened graph persistence against DoS (File Size Limits)

### PR DESCRIPTION
**Pull request by madmax983**
This PR addresses a potential Denial of Service (DoS) vulnerability where loading excessively large graph index or vector mapping files could lead to memory exhaustion or address space exhaustion.

Key changes:
- Enforced strict file size limits (`MAX_GRAPH_FILE_SIZE` = 4GB, `MAX_MAPPING_FILE_SIZE` = 1GB) when loading graph persistence files and HNSW mappings.
- Implemented a safe file loading pattern (`File::open` -> `metadata()` -> `read`) to prevent Time-of-Check Time-of-Use (TOCTOU) race conditions.
- Added new error variant `IndexPersistenceError::SizeLimitExceeded` for clear error reporting.
- Added comprehensive tests `test_load_oversized_graph_file` and `test_load_oversized_mapping_file`, utilizing `#[cfg(test)]` to lower limits (10MB) for practical verification.
- Updated the security journal (`.jules/warden.md`) with the new findings and defenses.

This hardening prevents an attacker from supplying a malicious (e.g., sparse) file to crash the server via OOM.

---
*PR created automatically by Jules for task [960064121218230737](https://jules.google.com/task/960064121218230737) started by @madmax983*

### CODE REVIEW

File: src/index/vector/hnsw.rs [hunk 1]
```diff
@@ -1100,7 +1106,33 @@ fn load_mappings_with_integrity(
...
+        )))
+    })?;
+
+    // Validate file size to prevent DoS
+    let metadata = file.metadata().map_err(|e| {
+        Error::Vector(VectorError::IndexError(format!(
+            "Failed to get mapping file metadata: {}",
+            e
+        )))
+    })?;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/hnsw.rs]: > medium
> 
> The error handling for I/O operations can be simplified. Instead of manually mapping the `io::Error` to a string-based `VectorError::IndexError`, you can use the `?` operator. The `crate::utils::error::Error` enum already has an `Io(io::Error)` variant and an `impl From<io::Error>`, so the conversion will be handled automatically. This makes the code more concise and preserves the original error type, which is better for debugging. This simplification can be applied to `File::open`, `file.metadata()`, and later `file.read_to_end()`.
> 
>     let mut file = std::fs::File::open(mappings_path)?;
> 
>     // Validate file size to prevent DoS
>     let metadata = file.metadata()?;

@jules

File: src/index/vector/hnsw.rs [hunk 2]
```diff
@@ -1100,7 +1106,33 @@ fn load_mappings_with_integrity(
...
+
+    if metadata.len() > MAX_MAPPING_FILE_SIZE {
+        return Err(Error::Vector(VectorError::IndexError(format!(
+            "Mapping file too large: {} bytes (limit: {} bytes)",
+            metadata.len(),
+            MAX_MAPPING_FILE_SIZE
+        ))));
+    }
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/hnsw.rs:1123]: > medium
> 
> For consistency with the changes in `graph.rs` and for clearer error handling, consider adding a `SizeLimitExceeded` variant to `VectorError` instead of using the generic `IndexError(String)`. This makes the error more specific and easier to handle programmatically by callers.
> 
> This would involve adding a variant to `VectorError` in `src/utils/error.rs`:
> 
> // in src/utils/error.rs
> SizeLimitExceeded {
>     message: String,
> },
> Then you could use it here. The test `test_load_oversized_mapping_file` would also need to be updated to match on the new error variant.
> 
>     if metadata.len() > MAX_MAPPING_FILE_SIZE {
>         return Err(Error::Vector(VectorError::SizeLimitExceeded {
>             message: format!(
>                 "Mapping file too large: {} bytes (limit: {} bytes)",
>                 metadata.len(),
>                 MAX_MAPPING_FILE_SIZE
>             ),
>         }));
>     }

@jules

[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/hnsw.rs:1123]: > medium
> 
> For consistency with the changes in `graph.rs` and for clearer error handling, consider adding a `SizeLimitExceeded` variant to `VectorError` instead of using the generic `IndexError(String)`. This makes the error more specific and easier to handle programmatically by callers.
> 
> This would involve adding a variant to `VectorError` in `src/utils/error.rs`:
> 
> // in src/utils/error.rs
> SizeLimitExceeded {
>     message: String,
> },
> Then you could use it here. The test `test_load_oversized_mapping_file` would also need to be updated to match on the new error variant.
> 
>     if metadata.len() > MAX_MAPPING_FILE_SIZE {
>         return Err(Error::Vector(VectorError::SizeLimitExceeded {
>             message: format!(
>                 "Mapping file too large: {} bytes (limit: {} bytes)",
>                 metadata.len(),
>                 MAX_MAPPING_FILE_SIZE
>             ),
>         }));
>     }

@jules

File: src/index/vector/hnsw.rs [hunk 3]
```diff
@@ -1100,7 +1106,33 @@ fn load_mappings_with_integrity(
...
+
+    // Read data
+    let mut file_data = Vec::with_capacity(metadata.len() as usize);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/index/vector/hnsw.rs]: > Potential overflow when casting u64 to usize on 32-bit systems. Although MAX_MAPPING_FILE_SIZE is 1GB which is safely within usize::MAX on 32-bit platforms, it's better to use safe conversion for consistency and to prevent future issues if the limit is ever increased.
> 
> Consider using try_into() with proper error handling:
> 
> let capacity = metadata.len().try_into().map_err(|_| {
>     Error::Vector(VectorError::IndexError(
>         format!("File size {} exceeds platform capacity", metadata.len())
>     ))
> })?;
> let mut file_data = Vec::with_capacity(capacity);
> Suggested change
>     let mut file_data = Vec::with_capacity(metadata.len() as usize);
>     let capacity: usize = metadata.len().try_into().map_err(|_| {
>         Error::Vector(VectorError::IndexError(format!(
>             "File size {} exceeds platform capacity",
>             metadata.len()
>         )))
>     })?;
>     let mut file_data = Vec::with_capacity(capacity);

@jules

File: src/storage/index_persistence/graph.rs [hunk 4]
```diff
@@ -159,7 +165,24 @@ pub fn save_graph_index(data: &GraphIndexData, path: &Path) -> Result<()> {
...
+                "Graph index file too large: {} bytes (limit: {} bytes)",
+                metadata.len(),
+                MAX_GRAPH_FILE_SIZE
+            ),
+        });
+    }
+
+    // Pre-allocate buffer based on file size
+    let mut bytes = Vec::with_capacity(metadata.len() as usize);
+    use std::io::Read;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/index_persistence/graph.rs]: > The inline import statement should be moved to the top of the file with other imports for better code organization and consistency. Currently, std::io::Read is imported inline in the function body at lines 184 and 618, but it would be cleaner to add it to the module-level imports at the top of the file.

@jules

File: src/storage/index_persistence/graph.rs [hunk 5]
```diff
@@ -562,8 +599,24 @@ pub fn load_graph_index_with_delta(base_path: &Path, delta_path: &Path) -> Resul
...
+                metadata.len(),
+                MAX_GRAPH_FILE_SIZE
+            ),
+        });
+    }
+
     // Load and decompress delta
-    let bytes = fs::read(delta_path)?;
+    let mut bytes = Vec::with_capacity(metadata.len() as usize);
+    use std::io::Read;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/index_persistence/graph.rs]: > The inline import statement should be moved to the top of the file with other imports for better code organization and consistency. Currently, std::io::Read is imported inline in the function body at lines 184 and 618, but it would be cleaner to add it to the module-level imports at the top of the file.

@jules

**Comment by madmax983** [comment]
> # Security Review: DoS Hardening for Graph Persistence
> ## Summary
> This PR adds critical DoS protections against memory/address space exhaustion by enforcing file size limits when loading graph index and vector mapping files. The implementation is **well-designed and follows security best practices**.
> 
> ## ✅ Strengths
> ### 1. **Security Implementation**
> * **TOCTOU Protection**: Correctly uses `File::open() -> metadata() -> read()` pattern to avoid race conditions
> * **Appropriate Limits**: 4GB for graph files, 1GB for mapping files are reasonable bounds
> * **Clear Error Messages**: New `SizeLimitExceeded` error variant provides helpful context
> * **Comprehensive Coverage**: All three load functions protected (`load_graph_index`, `load_graph_index_mmap`, `load_graph_index_with_delta`)
> 
> ### 2. **Code Quality**
> * **Consistent Pattern**: Same validation approach across all affected functions
> * **Pre-allocation Optimization**: `Vec::with_capacity(metadata.len())` reduces allocations
> * **Good Test Coverage**: Tests verify oversized file rejection with lowered limits
> * **Documentation**: Security journal updated with clear threat model and defenses
> 
> ### 3. **Testing Strategy**
> * **Practical Testing**: `#[cfg(test)]` reduces limits to 10MB for feasible tests
> * **Sparse File Usage**: `file.set_len()` avoids writing large test files
> * **Error Verification**: Tests confirm both error occurrence and message content
> 
> ## 🔍 Minor Observations
> ### 1. **Import Organization** (Line 184, 618)
> The `use std::io::Read` import appears inside functions rather than at the module level. While this works, consider moving to the top imports for consistency:
> 
> ```rust
> // At module level (line ~3)
> use std::fs;
> use std::io::Read; // ← Add here
> use std::path::Path;
> ```
> 
> **Impact**: Style/consistency only - no functional issue.
> 
> ### 2. **Test Assertion Specificity**
> In `test_load_oversized_mapping_file` (line 1988-1990), the test checks for the exact limit string. This creates tight coupling between test and implementation:
> 
> ```rust
> assert!(msg.contains("limit: 10485760 bytes")); // 10MB
> ```
> 
> **Suggestion**: Consider checking for the pattern rather than exact bytes:
> 
> ```rust
> assert!(msg.contains("limit:") && msg.contains("bytes"));
> ```
> 
> **Impact**: Low - makes tests more resilient to limit changes.
> 
> ### 3. **Consistency Opportunity**
> The `load_graph_index_mmap` function (line 343-352) could benefit from the same buffer pre-allocation comment pattern used in `load_graph_index`:
> 
> ```rust
> // Pre-allocate buffer based on file size
> let mut bytes = Vec::with_capacity(metadata.len() as usize);
> ```
> 
> **Impact**: Documentation clarity only.
> 
> ### 4. **Error Type Enhancement** (Optional)
> Consider adding structured fields to `SizeLimitExceeded`:
> 
> ```rust
> #[error("Size limit exceeded: {actual} bytes (limit: {limit} bytes)")]
> SizeLimitExceeded {
>     actual: u64,
>     limit: u64,
>     file_type: String, // "graph", "mapping", etc.
> },
> ```
> 
> **Benefit**: Easier programmatic error handling and metrics collection.
> 
> ## 🎯 Architecture Alignment
> ✅ **Follows CLAUDE.md Guidelines**:
> 
> * Uses strong typing (`IndexPersistenceError` enum)
> * No `.unwrap()` in production code
> * Comprehensive error handling
> * Security-first design
> * Performance-conscious (pre-allocation)
> * Well-documented threat model in `.jules/warden.md`
> 
> ✅ **Matches Existing Patterns**: The implementation mirrors existing DoS protections:
> 
> * `MAX_STRING_COUNT` / `MAX_STRING_LENGTH` (lines 122-129 in mod.rs)
> * `MAX_VECTOR_DIMENSIONS` (line 134 in mod.rs)
> * `MAX_SEARCH_RESULTS` in HNSW
> 
> ## 🚀 Performance Considerations
> **Positive Impact**:
> 
> * **Pre-allocation**: `Vec::with_capacity()` reduces allocations during file reads
> * **Early Validation**: File size check happens before expensive operations
> * **No Overhead for Valid Files**: Check is O(1) metadata read
> 
> **No Performance Regressions Expected**: The metadata check adds negligible latency (~μs) compared to file I/O.
> 
> ## 🔐 Security Assessment
> **Threat Model**: ✅ Correctly addresses unbounded file read DoS **Attack Vectors**: ✅ All covered (regular read, mmap, delta loading) **TOCTOU**: ✅ Properly mitigated with file descriptor reuse **Error Handling**: ✅ No panics, clear error messages **Test Coverage**: ✅ Comprehensive validation of edge cases
> 
> ## 📋 Recommendations
> ### Must-Have (Before Merge):
> None - the PR is ready to merge as-is.
> 
> ### Nice-to-Have (Future PRs):
> 1. Consider structured error fields for better observability
> 2. Move `use std::io::Read` to module-level imports for consistency
> 3. Add metrics/logging for size limit violations (for attack detection)
> 
> ## ✅ Final Verdict
> **LGTM** - This PR significantly improves the security posture of GallifreyDB with minimal code changes and excellent test coverage. The implementation follows best practices and aligns perfectly with the project's security-first philosophy.
> 
> **Recommended Actions**:
> 
> * ✅ Merge after CI passes
> * Consider backporting to stable branches if applicable
> * Monitor for size limit violations in production logs
> 
> _Review completed by Claude Code following [CLAUDE.md](../CLAUDE.md) security and quality standards_

@jules

**Comment by madmax983** [comment]
> # Security Review: DoS Hardening for Graph Persistence
> ## Summary
> This PR adds critical DoS protections against memory/address space exhaustion by enforcing file size limits when loading graph index and vector mapping files. The implementation is **well-designed and follows security best practices**.
> 
> ## ✅ Strengths
> ### 1. **Security Implementation**
> * **TOCTOU Protection**: Correctly uses `File::open() -> metadata() -> read()` pattern to avoid race conditions
> * **Appropriate Limits**: 4GB for graph files, 1GB for mapping files are reasonable bounds
> * **Clear Error Messages**: New `SizeLimitExceeded` error variant provides helpful context
> * **Comprehensive Coverage**: All three load functions protected (`load_graph_index`, `load_graph_index_mmap`, `load_graph_index_with_delta`)
> 
> ### 2. **Code Quality**
> * **Consistent Pattern**: Same validation approach across all affected functions
> * **Pre-allocation Optimization**: `Vec::with_capacity(metadata.len())` reduces allocations
> * **Good Test Coverage**: Tests verify oversized file rejection with lowered limits
> * **Documentation**: Security journal updated with clear threat model and defenses
> 
> ### 3. **Testing Strategy**
> * **Practical Testing**: `#[cfg(test)]` reduces limits to 10MB for feasible tests
> * **Sparse File Usage**: `file.set_len()` avoids writing large test files
> * **Error Verification**: Tests confirm both error occurrence and message content
> 
> ## 🔍 Minor Observations
> ### 1. **Import Organization** (Line 184, 618)
> The `use std::io::Read` import appears inside functions rather than at the module level. While this works, consider moving to the top imports for consistency:
> 
> ```rust
> // At module level (line ~3)
> use std::fs;
> use std::io::Read; // ← Add here
> use std::path::Path;
> ```
> 
> **Impact**: Style/consistency only - no functional issue.
> 
> ### 2. **Test Assertion Specificity**
> In `test_load_oversized_mapping_file` (line 1988-1990), the test checks for the exact limit string. This creates tight coupling between test and implementation:
> 
> ```rust
> assert!(msg.contains("limit: 10485760 bytes")); // 10MB
> ```
> 
> **Suggestion**: Consider checking for the pattern rather than exact bytes:
> 
> ```rust
> assert!(msg.contains("limit:") && msg.contains("bytes"));
> ```
> 
> **Impact**: Low - makes tests more resilient to limit changes.
> 
> ### 3. **Consistency Opportunity**
> The `load_graph_index_mmap` function (line 343-352) could benefit from the same buffer pre-allocation comment pattern used in `load_graph_index`:
> 
> ```rust
> // Pre-allocate buffer based on file size
> let mut bytes = Vec::with_capacity(metadata.len() as usize);
> ```
> 
> **Impact**: Documentation clarity only.
> 
> ### 4. **Error Type Enhancement** (Optional)
> Consider adding structured fields to `SizeLimitExceeded`:
> 
> ```rust
> #[error("Size limit exceeded: {actual} bytes (limit: {limit} bytes)")]
> SizeLimitExceeded {
>     actual: u64,
>     limit: u64,
>     file_type: String, // "graph", "mapping", etc.
> },
> ```
> 
> **Benefit**: Easier programmatic error handling and metrics collection.
> 
> ## 🎯 Architecture Alignment
> ✅ **Follows CLAUDE.md Guidelines**:
> 
> * Uses strong typing (`IndexPersistenceError` enum)
> * No `.unwrap()` in production code
> * Comprehensive error handling
> * Security-first design
> * Performance-conscious (pre-allocation)
> * Well-documented threat model in `.jules/warden.md`
> 
> ✅ **Matches Existing Patterns**: The implementation mirrors existing DoS protections:
> 
> * `MAX_STRING_COUNT` / `MAX_STRING_LENGTH` (lines 122-129 in mod.rs)
> * `MAX_VECTOR_DIMENSIONS` (line 134 in mod.rs)
> * `MAX_SEARCH_RESULTS` in HNSW
> 
> ## 🚀 Performance Considerations
> **Positive Impact**:
> 
> * **Pre-allocation**: `Vec::with_capacity()` reduces allocations during file reads
> * **Early Validation**: File size check happens before expensive operations
> * **No Overhead for Valid Files**: Check is O(1) metadata read
> 
> **No Performance Regressions Expected**: The metadata check adds negligible latency (~μs) compared to file I/O.
> 
> ## 🔐 Security Assessment
> **Threat Model**: ✅ Correctly addresses unbounded file read DoS **Attack Vectors**: ✅ All covered (regular read, mmap, delta loading) **TOCTOU**: ✅ Properly mitigated with file descriptor reuse **Error Handling**: ✅ No panics, clear error messages **Test Coverage**: ✅ Comprehensive validation of edge cases
> 
> ## 📋 Recommendations
> ### Must-Have (Before Merge):
> None - the PR is ready to merge as-is.
> 
> ### Nice-to-Have (Future PRs):
> 1. Consider structured error fields for better observability
> 2. Move `use std::io::Read` to module-level imports for consistency
> 3. Add metrics/logging for size limit violations (for attack detection)
> 
> ## ✅ Final Verdict
> **LGTM** - This PR significantly improves the security posture of GallifreyDB with minimal code changes and excellent test coverage. The implementation follows best practices and aligns perfectly with the project's security-first philosophy.
> 
> **Recommended Actions**:
> 
> * ✅ Merge after CI passes
> * Consider backporting to stable branches if applicable
> * Monitor for size limit violations in production logs
> 
> _Review completed by Claude Code following [CLAUDE.md](../CLAUDE.md) security and quality standards_

@jules
